(trunk) PXB-3279 Remove requirement to have open file handles equal to files …

aybek opened the pull request:
…in datadir

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1641) <br/>All committers have signed the CLA.

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3210,6 +3212,9 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
     goto skip;
+  } else if (res == XB_FIL_CUR_MISSING) {
+    ddl_tracker->add_missing_after_discovery(cursor.space_id);
+    goto error;
```
shouldn't it be continue here? if a file is missing, we should continue

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3248,9 +3253,10 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
+    xb::info() << action << " " << node->space->id << " " << node_path;
   } else {
-    xb::info() << action << " " << node_path << " to " << dstfile->path;
+    xb::info() << action << " " << node->space->id << " " << node_path << " to "
```
May be it would be better something like this: action <filename> with space_id  10;

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3273,10 +3279,11 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
-    xb::info() << "Done: " << action << " " << node_path << " to "
-               << dstfile->path;
+    xb::info() << "Done: " << action << " " << node->space->id << " "
+               << node_path << " to " << dstfile->path;
```
same as above

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3366,15 +3373,29 @@ static void data_copy_thread_func(data_thread_ctxt_t *ctxt) {
+      // failure
+      if (ddl_tracker != nullptr && opt_lock_ddl == LOCK_DDL_REDUCED) {
+        // ingore failure and continue, we should ideally track these..
+        continue;
```
Please treat this as a failure and abort backup. I think XB_FIL_CUR_MISSING is designed to handle all known issues with copying files during reduced lock.

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/reducedlock/limit_file_handles.sh:
```diff
@@ -0,0 +1,74 @@
+require_debug_pxb_version
+require_pro_pxb_version
+
+open_file_limit=$(ulimit -n)
+ulimit -n 1024
```
Did you run the test on Jenkins? The user might not have permission to execute this.

satya-bodapati replied:
For now keep it, we will see when jenkins works.

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/reducedlock/limit_file_handles.sh:
```diff
@@ -0,0 +1,74 @@
+
+innodb_wait_for_flush_all
+
+xtrabackup --backup --target-dir=$topdir/backup \
+  --debug-sync="xtrabackup_suspend_between_file_discovery_and_open" --lock-ddl=REDUCED \
+  2> >( tee $topdir/backup.log)&
```
We should also test dropping and renaming files, after discovery & loading to cache. Can you please add this test scenarios?

aybek replied:
do you mean after sync point `xtrabackup_load_tablespaces_pause` ?
I think this scenario is already covered, for instance in
/mnt/nvme0_xfs/dev_aibekb/pxb/backup_pxb3279/storage/innobase/xtrabackup/test/suites/reducedlock/basic_operation.sh

satya-bodapati replied:
Can you please ensure that the file opening fails from your new code? XB_FIL_CUR_MISSING is returned.

Also do we have testcase for space_id mismatch? This is to ensure the space_id mismatch path is activated/used.

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -145,16 +145,14 @@ void ddl_tracker_t::add_to_recopy_tables(space_id_t space_id, lsn_t record_lsn,
-  missing_after_discovery.insert(path);
+  missing_after_discovery.insert(space_id);
+
+  xb::info() << " missing space ID: " << space_id;
```
Let's add the same prefix DDL tracking: missing after discovery space_id :

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -593,6 +589,8 @@ dberr_t ddl_tracker_t::handle_ddl_operations() {
       continue;
     }
 
+    ulint flags = tables_copied_no_lock[space_id].second;
```
Lets not use the [] notation anymore. Please use the find() method.

satya-bodapati commented on storage/innobase/xtrabackup/src/fil_cur.cc:
```diff
@@ -151,7 +151,7 @@ xb_fil_cur_result_t xb_fil_cur_open(
 
       xb::error() << "cannot open tablespace " << cursor->abs_path;
 
-      return (XB_FIL_CUR_ERROR);
+      return (XB_FIL_CUR_MISSING);
```
Should be we keep XB_FIL_CUR_ERROR here? because this path, the caller doesn't need to close the file.

May be better to simplify. Let the caller always call the close file function? I think it is OK. because we have stored the state FIL_CLOSED or OPENED)

aybek replied:
Do you think we need to keep the XB_FIL_CUR_ERROR here ?

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3210,6 +3212,9 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
   res = xb_fil_cur_open(&cursor, read_filter, node, thread_n);
   if (res == XB_FIL_CUR_SKIP) {
     goto skip;
+  } else if (res == XB_FIL_CUR_MISSING) {
+    ddl_tracker->add_missing_after_discovery(cursor.space_id);
+    goto skip;
```
do we call this function under the second copy phase? (ie under the lock)?

After taking the lock, any error is a serious error, and we should abort the backup.

aybek replied:
I will check this

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3248,9 +3253,11 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
 
   if (xtrabackup_stream) {
-    xb::info() << action << " " << node_path;
+    xb::info() << action << " with space_id " << node->space->id << " "
```
Please change to " file with space_id "

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3248,9 +3253,11 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
+               << node_path;
   } else {
-    xb::info() << action << " " << node_path << " to " << dstfile->path;
+    xb::info() << action << " with space_id " << node->space->id << " "
```
file with space_id

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3273,10 +3280,11 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
   /* close */
   if (xtrabackup_stream) {
-    xb::info() << "Done: " << action << " " << node_path;
+    xb::info() << "Done: " << action << " " << node->space->id << " "
```
fix here to; "file with space_id "

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3273,10 +3280,11 @@ bool xtrabackup_copy_datafile_func(fil_node_t *node, uint thread_n,
   } else {
-    xb::info() << "Done: " << action << " " << node_path << " to "
-               << dstfile->path;
+    xb::info() << "Done: " << action << " with space_id " << node->space->id
```
file with space_id

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -3366,16 +3374,25 @@ static void data_copy_thread_func(data_thread_ctxt_t *ctxt) {
+          ddl_tracker->add_table_from_ibd_scan(node->space->id,
+                                               node->space->files.front().name,
+                                               node->space->flags);
+        }
```
I would an else branch. this will be unhandled path? On release build, print message about space->id, name and return error.

On debug build, let it crash ut_ad(0);

aybek replied:
Do you mean we need to crash in `else case` (not fsp_is_undo_tablespace and not fsp_is_ibd_tablespace) ?

satya-bodapati replied:
yes, we should hanlde all cases So am talking about the else branch
if () {
} else if () {
} else {
--> unexpected path, so crash here on debug builds and  print warning for releae builds
}

aybek replied:
left this as is, in else case we have all other tablespaces like system/tmp etc

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -4083,80 +4100,6 @@ static uint xb_set_max_open_files(
 }
 
-/* true on success, false on failure */
-bool xb_check_and_set_open_files_limit(size_t num_files) {
```
can you make this as separate commit, remove "commit id"? use `git revert`

aybek replied:
commited separately but without revert, all reducedlock feature commits were squashed during merge

satya-bodapati commented on storage/innobase/xtrabackup/test/suites/reducedlock/rename_problem.sh:
```diff
@@ -0,0 +1,54 @@
+  fi
+
+  xtrabackup --prepare --target-dir=$topdir/backup_enc_general_tablespace --xtrabackup-plugin-dir=${plugin_dir}
+#
+#
+#  stop_server
+#
+#  rm -rf $mysql_datadir/*
+#  xtrabackup --copy-back --target-dir=$topdir/backup_enc_general_tablespace --xtrabackup-plugin-dir=${plugin_dir}
+#
+#  start_server
+#  verify_db_state test
+#  stop_server
+#  rm -rf $mysql_datadir $topdir/backup_with_enc_general_tablespace.log $topdir/backup_enc_general_tablespace $topdir/backup_inc
```
I think these parts should be activated. Wihtout the patch, the prepare would crash. But since we have a fix, we have to restore, start_server, etc.

satya-bodapati commented:
@aybek can you also please rebase the patch to 8.4 and also create a trunk version of it?

aybek committed:
Fix rename bug
This covers a case when we have 2 RENAME DDLs on the same table,
first rename happens before the files are discovered
second rename happens after files discovered but before files are opened for loading to cache

aybek committed:
Reverting PXB-3381 Implement check to fail early if number of open file handles is not same as number of files in datadir

aybek committed:
PXB-3279 Remove requirement to have open file handles equal to files in datadir

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -548,6 +556,12 @@ dberr_t ddl_tracker_t::handle_ddl_operations() {
     }
   }
 
+  for (auto &table : missing_after_discovery) {
```
@aybek Do you know why need this? We will not add any table which is missing to tables_copied_no_lock. isn't it? :thinking:

satya-bodapati replied:
Fixed by the new commit

satya-bodapati committed:
PXB-3279 : Remove requirement to have open file handles equal to files in datadir

Additional fixes:
-----------------
1. missing files are tracked in copied_files_no_lock. They are now
   tracked properly. So only copied files are tracked in
   tables_copied_no_lock map

2. Corrupted tablespaces and recopy of it assumed the table is always
   in tables_copied_no_lock map. this is not the case anymore after the
   fix 1 above. Additional code to handle this

3. get_space_flags() worked on tables in tables_copied_no_lock map. For
   undo, it fails with debug assertion. Undo tablespaces space_flags are
   always returned as 0. We always create undo tablespaces in top
   datadir (not in database dirs)

4. Suppress ERROR that are expected because of missing files. this can
   happen because DDLs can remove or rename files without the backup
   lock. After server is locked, such errors are not tolerated.

5. Additional logging

6. Testcase fix

satya-bodapati committed:
Fix unused variable

satya-bodapati committed:
testcase fix. Lower the number to tables to reproduce open files limit issue. With 5000 tables, it times out. 1500 is sufficient with 1024 handles limit

satya-bodapati commented:
#1664 handles this
